File: warning.py
import pandas as pd
import numpy as np
from scipy import stats
from datetime import datetime

EXCEL_PATH = 'C:\\Users\\ennin\\Desktop\\work\\Documents\\新车落地服务20180907 (1).xls'
NEW_PATH = 'C:\\Users\\ennin\\Desktop\\work\\Documents\\test_data2.xlsx'
NODE_NAMES = ['订单匹配时间', '通知出库时间', '开票通知时间', '实际出库时间', '开票时间', '发合寄件时间', '交强险购买时间', '商业险购买时间',
'交强险寄件时间', '购置税申请时间', '购置税打款时间', '发合收件时间', '购置税收款时间', '接车工单完成时间', '交强险收件时间']
INDEX_DATA = [1, 2, 3, 4, 5, 15, 8, 9, 10, 14, 12, 11, 13]
LAST_NODES = [9, 10, 11, 12]
NEED_STRIP = [1, 3, 4]

# route index
path_index_0 = {0, 1, 3, 5, 6, 9}
path_index_1 = {0, 1, 3, 4, 7, 8, 10}
path_index_2 = {0, 1, 3, 4, 11}
path_index_3 = {0, 2, 12}
path_index = [path_index_0, path_index_1, path_index_2, path_index_3]

# route distance
path_time_0 = [1.44, 1.3, 0.93, 0.31, 1.55]
path_time_1 = [1.44, 1.3, 1.18, 1.01, 0.3, 0.66]
path_time_2 = [1.44, 1.3, 1.18, 1.9]
path_time_3 = [1.84, 3.62]
path_time = [path_time_0, path_time_1, path_time_2, path_time_3]

# order
path_list_0 = [0, 1, 3, 5, 6, 9]
path_list_1 = [0, 1, 3, 4, 7, 8, 10]
path_list_2 = [0, 1, 3, 4, 11]
path_list_3 = [0, 2, 12]
path_list = [path_list_0, path_list_1, path_list_2, path_list_3]


path_column_0 = ['L_2', 'L_3', 'L_9', 'L_10', 'L_11']
path_column_1 = ['L_2', 'L_3', 'L_4', 'L_6', 'L_7', 'L_8']
path_column_2 = ['L_2', 'L_3', 'L_4', 'L_5']
path_column_3 = ['L_0', 'L_1']
path_column = [path_column_0, path_column_1, path_column_2, path_column_3]

normal_data = pd.read_csv('normal_data.csv')


def delay_probability(path_list_index, warning_node_index, now_time, node_time, early_time):
    select_path_column = path_column[path_list_index]
    current_l = select_path_column[warning_node_index - 1]
    remain_l = select_path_column[warning_node_index:]
    current_l_df = normal_data[current_l]/24
    remain_l_df = normal_data[remain_l]/24
    past_time = now_time - early_time
    remain_time_max = 3 - past_time
    node_past_time = now_time - node_time
    current_l_df = current_l_df - node_past_time
    current_l_df = np.maximum(current_l_df, 0)
    sum_df = current_l_df + remain_l_df.T.sum()
    if remain_time_max <= 0:
        probability = 1
    else:
        z_score = (remain_time_max - sum_df.mean()) / sum_df.std()
        probability = 1 - stats.norm.cdf(z_score)
    return probability


class Stack(object):
    # 初始化栈为空列表
    def __init__(self):
        self.items = []

    # 判断栈是否为空，返回布尔值
    def is_empty(self):
        return self.items == []

    # 返回栈顶元素
    def peek(self):
        return self.items[len(self.items) - 1]

    # 返回栈的大小
    def size(self):
        return len(self.items)

    # 把新的元素堆进栈里面（程序员喜欢把这个过程叫做压栈，入栈，进栈……）
    def push(self, item):
        self.items.append(item)

    # 把栈顶元素丢出去（程序员喜欢把这个过程叫做出栈……）
    def pop(self):
        return self.items.pop()


def load_data(data_path, sheet_name):
    raw_data = pd.read_excel(data_path, sheet_name=sheet_name)
    return raw_data


def warning(sorted_node, property_stack, node_times, start_time):
    result_set = set()
    tmp_set = set()
    # processing_nodes = []
    # 判断有没有路径完成
    not_finished = True
    # 记录首次到达的节点
    first_node = -1
    for node in sorted_node:
        # 判断当前node 有没有被移除
        removed = False
        while not removed:
            # 当告警的节点已经响应，从告警池中删除
            if node in result_set:
                result_set.remove(node)
            # 当前node 在栈顶集合
            if not property_stack.is_empty() and node in property_stack.peek():
                property_stack.peek().remove(node)
                removed = True
                # 判断是否有路径完成
                if not_finished and node in LAST_NODES:
                    not_finished = False
                    first_node = node

            # tmp_set 用于存放告警节点
            elif node not in tmp_set:
                pop_set = property_stack.pop()
                result_set = result_set | pop_set
                tmp_set = tmp_set | pop_set

            else:
                removed = True
            if not property_stack.is_empty() and not property_stack.peek():
                property_stack.pop()

    if not not_finished:
        # 获取当前告警节点
        curr_warning = filter_result_fixed(result_set)
        tmp_warning = curr_warning.copy()
        for warning_node in curr_warning:
            # 遍历各个路径
            tmp_remain_time_list = []
            remove_warning = True
            for path_list_index, path_list_ in enumerate(path_list):
                if warning_node in path_list_:
                    # 获取当前节点索引
                    warning_node_index = path_list_.index(warning_node)
                    ########### plan A ###########
                    # 差值
                    # node_times[sorted_node[-1]] todo 需要用now替换
                    delta = path_time[path_list_index][warning_node_index - 1] - (
                                    node_times[sorted_node[-1]] - node_times[path_list_[warning_node_index - 1]])
                    part_sum = sum(path_time[path_list_index][warning_node_index:])
                    delta_2 = node_times[sorted_node[-1]] - node_times[first_node]
                    tmp_remain_time = 3 - (max(0, delta) + part_sum + delta_2)
                    tmp_remain_time_list.append(tmp_remain_time)
                    ########### plan B ###########
                    # # node_times[sorted_node[-1]] todo 需要用now替换
                    probability_ = delay_probability(path_list_index, warning_node_index, node_times[sorted_node[-1]],
                                                     node_times[path_list_[warning_node_index - 1]], node_times[first_node])
                    print('path: {}, node num: {}, probability: {}.'.format(path_list_index, warning_node, probability_))
            if tmp_remain_time_list:
                for data_ in tmp_remain_time_list:
                    if data_ < 0:
                        remove_warning = False
                if remove_warning:
                    tmp_warning.remove(warning_node)

    else:
        tmp_warning = filter_result_fixed(result_set)

    return tmp_set, result_set, tmp_warning


# 过滤初步告警节点，保留每条路径上优先级最高的节点
def filter_result_fixed(result):
    last_node = set()
    for path in path_index:
        tmp_set = result & path
        if len(tmp_set) > 1:
            tmp_min = 9
            for i_node in tmp_set:
                if i_node < tmp_min:
                    tmp_min = i_node
            last_node = last_node | {tmp_min}
        else:
            last_node = last_node | tmp_set
    return last_node


def init_stack():
    property_1 = {0}
    property_2 = {1, 2, 3}
    property_3 = {4, 5, 6}
    property_4 = {7, 8}
    property_5 = {9, 10, 11, 12}
    property_stack = Stack()
    property_stack.push(property_5)
    property_stack.push(property_4)
    property_stack.push(property_3)
    property_stack.push(property_2)
    property_stack.push(property_1)
    return property_stack


def process_raw_data(one_line_data):
    # 1. strip [0, 2, 3]
    for part_data_01_index in NEED_STRIP:
        part_data_01 = one_line_data[part_data_01_index]
        if part_data_01 is not np.nan:
            one_line_data[part_data_01_index] = strip_data(part_data_01)

    # # 1
    # for index, data_ in enumerate(one_line_data):
    #     if data_ is not np.nan and len(data_) != 19:
    #         one_line_data[index] = strip_data(data_)

    # 2. convert str time to datetime time
    for index, data_ in enumerate(one_line_data):
        one_line_data[index] = convert_to_datetime(data_)

    # 3. calculate diff time
    start_time = one_line_data[0]
    for index, data_ in enumerate(one_line_data):
        one_line_data[index] = diff_time(data_, start_time)

    # 4. 计算车险购买时间
    one_line_data = np.append(one_line_data, max(one_line_data[6], one_line_data[7]))

    # 5. finished
    processed_data = one_line_data[INDEX_DATA]
    return processed_data, start_time


# 截取时间尾部的.0
# eg. 2018-08-10 12:17:13.0 to 2018-08-10 12:17:13
def strip_data(tar):
    tar = str(tar)
    return tar[:19]


# 将str类型的time数据转化为datetime类型的数据
def convert_to_datetime(str_data):
    if str_data is not np.nan:
        str_data = datetime.strptime(str_data, '%Y-%m-%d %H:%M:%S')
    return str_data


# 计算diff time
def diff_time(date_time, start):
    if date_time is not np.nan:
        date_time -= start
        date_time = round(date_time.total_seconds() / 60 / 60 / 24, 2)
    return date_time


# 排序
def sort_index(node_times_):
    node_times_pd = pd.Series(node_times_)
    arrived = len(node_times_pd.dropna())
    sorted_arg = np.argsort(node_times_)
    sorted_index = sorted_arg[:arrived]
    return sorted_index


# convert now to datetime
def convert_now(start_time):
    now = datetime.now()
    now_strip = strip_data(str(now))
    now = datetime.strptime(now_strip, '%Y-%m-%d %H:%M:%S')
    # start_time = datetime.strptime(start_time, '%Y-%m-%d %H:%M:%S')
    diff_time_ = now - start_time
    diff_time_ = round(diff_time_.total_seconds() / 60 / 60 / 24, 2)
    return diff_time_


def single_data_test(single_data):
    processed_data_ = single_data
    start_time_ = ''
    need_index = sort_index(processed_data_)
    property_stack_ = init_stack()
    tmp_set_, result_set_, tmp_warning_ = warning(need_index, property_stack_, processed_data_, start_time_)
    return tmp_set_, result_set_, tmp_warning_


def single_data_test_(single_data_):
    processed_data_, start_time_ = process_raw_data(single_data_)
    need_index = sort_index(list(processed_data_))
    property_stack_ = init_stack()
    tmp_set_, result_set_, tmp_warning_ = warning(need_index, property_stack_, processed_data_, start_time_)
    return processed_data_, tmp_set_, result_set_, tmp_warning_


def main():
    # process raw data
    excel_data = load_data(EXCEL_PATH, 'Sheet1')
    need_data = excel_data[NODE_NAMES]

    for index, data_ in enumerate(need_data.values):
        processed_data_a, tmp_set_a, result_set_a, tmp_warning_a = single_data_test_(data_)
        if 0 not in tmp_set_a and tmp_warning_a:
            print(index)
            print(processed_data_a)
            print(tmp_set_a)
            print(result_set_a)
            print(tmp_warning_a)
            print('-----------------------------------------')


def main_():
    single_data_a = [4.92, 5.78, np.nan, 5.81, np.nan, 7.07, 7.55, np.nan, np.nan, np.nan, np.nan, np.nan, np.nan]
    tmp_set_b, result_set_b, tmp_warning_b = single_data_test(single_data_a)
    print(tmp_set_b)
    print(result_set_b)
    print(tmp_warning_b)


if __name__ == '__main__':
    main()

File: Warning.java
package com.souche;

import org.apache.commons.csv.CSVFormat;
import org.apache.commons.csv.CSVRecord;
import org.apache.commons.math3.distribution.NormalDistribution;
import org.apache.commons.math3.stat.descriptive.moment.Mean;
import org.apache.commons.math3.stat.descriptive.moment.StandardDeviation;

import java.io.Reader;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.*;

public class Warning {

    // 路径的叶子节点列表
    private static final List<Integer> LAST_NODES = new ArrayList<Integer>(Arrays.asList(9, 10, 11, 12));

    // 各个路径顺序节点列表
    private static final List<Integer> pathList_0 = new ArrayList<Integer>(Arrays.asList(0, 1, 3, 5, 6, 9));

    private static final List<Integer> pathList_1 = new ArrayList<Integer>(Arrays.asList(0, 1, 3, 4, 7, 8, 10));

    private static final List<Integer> pathList_2 = new ArrayList<Integer>(Arrays.asList(0, 1, 3, 4, 11));

    private static final List<Integer> pathList_3 = new ArrayList<Integer>(Arrays.asList(0, 2, 12));

    private static final List<List<Integer>> pathList = new ArrayList<List<Integer>>(Arrays.asList(pathList_0, pathList_1, pathList_2, pathList_3));

    // 各个路径顺序节点到达时间列表
    private static final List<Double> pathTime_0 = new ArrayList<Double>(Arrays.asList(1.44, 1.3, 0.93, 0.31, 1.55));

    private static final List<Double> pathTime_1 = new ArrayList<Double>(Arrays.asList(1.44, 1.3, 1.18, 1.01, 0.3, 0.66));

    private static final List<Double> pathTime_2 = new ArrayList<Double>(Arrays.asList(1.44, 1.3, 1.18, 1.9));

    private static final List<Double> pathTime_3 = new ArrayList<Double>(Arrays.asList(1.84, 3.62));

    private static final List<List<Double>> pathTime = new ArrayList<List<Double>>(Arrays.asList(pathTime_0, pathTime_1, pathTime_2, pathTime_3));

    // 用于计算概率
    private static final List<String> pathColumn_0 = new ArrayList<String>(Arrays.asList("L_2", "L_3", "L_9", "L_10", "L_11"));

    private static final List<String> pathColumn_1 = new ArrayList<String>(Arrays.asList("L_2", "L_3", "L_4", "L_6", "L_7", "L_8"));

    private static final List<String> pathColumn_2 = new ArrayList<String>(Arrays.asList("L_2", "L_3", "L_4", "L_5"));

    private static final List<String> pathColumn_3 = new ArrayList<String>(Arrays.asList("L_0", "L_1"));

    private static final List<List<String>> pathColumn = new ArrayList<List<String>>(Arrays.asList(pathColumn_0, pathColumn_1, pathColumn_2, pathColumn_3));

    // 缓存的 csv数据用于计算概率
    private static final String PATH = "C:\\Users\\ennin\\Desktop\\work\\Codes\\warnning_\\data\\normal_data.csv";

    // 路径节点集合
    private static final Set<Integer> pathIndex_0 = new HashSet<Integer>(){
        {
            add(0);
            add(1);
            add(3);
            add(5);
            add(6);
            add(9);
        }
    };

    private static final Set<Integer> pathIndex_1 = new HashSet<Integer>(){
        {
            add(0);
            add(1);
            add(3);
            add(4);
            add(7);
            add(8);
            add(10);
        }
    };

    private static final Set<Integer> pathIndex_2 = new HashSet<Integer>(){
        {
            add(0);
            add(1);
            add(3);
            add(4);
            add(11);
        }
    };

    private static final Set<Integer> pathIndex_3 = new HashSet<Integer>(){
        {
            add(0);
            add(2);
            add(12);
        }
    };

    // 各个路径节点集合的列表
    private static final List<Set<Integer>> pathIndex =
            new ArrayList<Set<Integer>>(Arrays.asList(pathIndex_0, pathIndex_1, pathIndex_2, pathIndex_3));
    

    // 初始化优先级栈
    public static Stack<Set<Integer>> initStack(){
        Stack<Set<Integer>> propertyStack = new Stack<Set<Integer>>();
        Set<Integer> property_1 = new HashSet<Integer>(){{
            add(0);
        }};

        Set<Integer> property_2 = new HashSet<Integer>(){{
            add(1);
            add(2);
            add(3);
        }};

        Set<Integer> property_3 = new HashSet<Integer>(){{
            add(4);
            add(5);
            add(6);
        }};

        Set<Integer> property_4 = new HashSet<Integer>(){{
            add(7);
            add(8);
        }};

        Set<Integer> property_5 = new HashSet<Integer>(){{
            add(9);
            add(10);
            add(11);
            add(12);
        }};

        propertyStack.push(property_5);
        propertyStack.push(property_4);
        propertyStack.push(property_3);
        propertyStack.push(property_2);
        propertyStack.push(property_1);
        return propertyStack;
    }

    // 过滤初步警报节点，保留每条路径上优先级最高的节点
    public static Set<Integer> filterResult(Set<Integer> result){
        Set<Integer> returnSet = new HashSet<Integer>();
        Set<Integer> tmpSet = new HashSet<Integer>();
        Set<Integer> tmpMinSet = new HashSet<Integer>();
        for(Set<Integer> path: pathIndex){
            tmpSet.clear();
            tmpSet.addAll(result);
            tmpSet.retainAll(path);
            if(tmpSet.size() > 1){
                int tmpMin = 9;
                for(int iNode: tmpSet){
                    if(iNode < tmpMin){
                        tmpMin = iNode;
                    }
                }
                tmpMinSet.clear();
                tmpMinSet.add(tmpMin);
                returnSet.addAll(tmpMinSet);
            }else {
                returnSet.addAll(tmpSet);
            }
        }
        return returnSet;
    }


    // 对到达时间列表进行排序，返回排序后节点索引
    public static int[] argsort(final double[] a, final boolean ascending) {
        int count = countNan(a);
        Integer[] indexes = new Integer[a.length];
        for (int i = 0; i < indexes.length; i++) {
            indexes[i] = i;
        }
        Arrays.sort(indexes, new Comparator<Integer>() {
            public int compare(final Integer i1, final Integer i2) {
                return (ascending ? 1 : -1) * Double.compare(a[i1], a[i2]);
            }
        });
        return asArray(count, indexes);
    }

    public static <T extends Number> int[] asArray(int count, final T... a) {
        int[] b = new int[count];
        for (int i = 0; i < b.length; i++) {
            b[i] = a[i].intValue();
        }
        return b;
    }

    public static int countNan(double[] oneLineData){
        int count = 0;
        for (double i : oneLineData){
            if (!Double.isNaN(i)){
                ++count;
            }
        }
        return count;
    }

    // 获取当前警告节点⚠
    public static Set<Integer> warning(int[] sortedNode, Stack<Set<Integer>> propertyStack, double[] nodeTimes, String startTime) throws Exception{
        Set<Integer> resultSet = new HashSet<Integer>();
        Set<Integer> tmp_set = new HashSet<Integer>();

        // 用于判断有没有路径完成
        boolean notFinished = true;
        // 用于记录首次到达的节点
        int firstNode = -1;
        for (int node : sortedNode){
            // 判断当前node 有没有被移除
            boolean removed = false;
            while (!removed){
                // 当报警节点已经响应，从报警集合中删除
                if (resultSet.contains(node)){
                    resultSet.remove(node);
                }
                // 判断当前node 是否在栈顶集合
                if (!propertyStack.isEmpty() && propertyStack.peek().contains(node)){
                    propertyStack.peek().remove(node);
                    removed = true;
                    // 判断移除的节点是否是路径叶子节点
                    if (notFinished && LAST_NODES.contains(node)){
                        notFinished = false;
                        firstNode = node;
                    }
                }
                // tmp_set 用于存放报警节点
                else if (!tmp_set.contains(node)){
                    Set<Integer> popSet = propertyStack.pop();
                    resultSet.addAll(popSet);
                    tmp_set.addAll(popSet);
                }
                else {
                    removed = true;
                }
                // 当栈不为空，且栈顶集合为空，移除空集合
                if (!propertyStack.isEmpty() && propertyStack.peek().isEmpty() ){
                    propertyStack.pop();
                }
            }
        }
        Set<Integer> resultWarning = filterResult(resultSet);
        if (!notFinished){
            // 获取当前报警节点集合
            Set<Integer> currWarning = filterResult(resultSet);
            for (int warningNode : currWarning){
                // 遍历各个路径
                List<Double> remainTimeList = new ArrayList<Double>();
                boolean removeWarning = true;
                int pathListIndex = 0;
                for (List<Integer> pathList_ : pathList){
                    if (pathList_.contains(warningNode)){
                        // 获取当前节点索引
                        int warningNodeIndex = pathList_.indexOf(warningNode);
                        // plan A
                        // 差值
                        // nodeTimes[sortedNode[sortedNode.length-1]] todo 需要用now替换
                        double delta = pathTime.get(pathListIndex).get(warningNodeIndex-1) -
                                (nodeTimes[sortedNode[sortedNode.length-1]] - nodeTimes[pathList_.get(warningNodeIndex-1)]);
                        double partSum = 0;
                        int partTimeLen = pathTime.get(pathListIndex).size();
                        for (int i = warningNodeIndex; i < partTimeLen; ++i){
                            partSum += pathTime.get(pathListIndex).get(i);
                        }
                        double delta_2 = nodeTimes[sortedNode[sortedNode.length-1]] - nodeTimes[firstNode];
                        double remainTime = 3 - (Math.max(0, delta) + partSum + delta_2);
                        remainTimeList.add(remainTime);
                        // plan B
                        // nodeTimes[sortedNode[sortedNode.length-1]] todo 需要用now替换
                        double probability_ = delayProbability(pathListIndex, warningNodeIndex, nodeTimes[sortedNode[sortedNode.length-1]],
                                nodeTimes[pathList_.get(warningNodeIndex-1)], nodeTimes[firstNode]);

                        System.out.println("Path: " + pathListIndex + ", node num: " + warningNode + ", probability: " + probability_);

                    }
                    ++ pathListIndex;
                }

                if (!remainTimeList.isEmpty()){
                    for (double time_ : remainTimeList){
                        if (time_ < 0){
                            removeWarning = false;
                        }
                    }
                    if (removeWarning){
                        resultWarning.remove(warningNode);
                    }
                }
            }
        }
        System.out.print("----------------------------------");
        return resultWarning;
    }

    public static double delayProbability (int pathListIndex, int warningNodeIndex, double nowTime, double nodeTime, double earlyTime) throws Exception{
        double probability = -1.0;
        Reader reader = Files.newBufferedReader(Paths.get(PATH));
        Iterable<CSVRecord> records = CSVFormat.RFC4180.withFirstRecordAsHeader().parse(reader);

        List<String> selectPathColumn = pathColumn.get(pathListIndex);
        String current_l = selectPathColumn.get(warningNodeIndex - 1);
        List<String> remain_l = new ArrayList<>();
        for (int i = warningNodeIndex; i < selectPathColumn.size(); ++i){
            remain_l.add(selectPathColumn.get(i));
        }

        double pastTime = nowTime - earlyTime;
        double remainTimeMax = 3 - pastTime;
        double nodePastTime = nowTime - nodeTime;

        List<Double> current_l_df = new ArrayList<>();
        // Map<String, Double> remain_l_df = new HashMap<>();
        List<Double> resultList = new ArrayList<>();
        List<Double> sumDf = new ArrayList<>();

        int tmpCount_1 = 0;
        for (CSVRecord record : records) {
            String L_0 = record.get(current_l);
            double tmpData_1 = Double.parseDouble(L_0) / 24;
            tmpData_1 -= nodePastTime;
            tmpData_1 = Math.max(0, tmpData_1);
            current_l_df.add(tmpData_1);

            int tmpCount_2 = 0;
            double tmpSum = 0;
            for (String tmpName: remain_l){
                String tmpData = record.get(tmpName);
                tmpSum += Double.parseDouble(tmpData) / 24;
                if (tmpCount_2 == remain_l.size() - 1){
                    resultList.add(tmpSum);
                }
                ++ tmpCount_2;
            }
            sumDf.add(tmpData_1 + resultList.get(tmpCount_1));
            ++ tmpCount_1;
        }


        double[] sumDfDouble = new double[sumDf.size()];
        StandardDeviation std = new StandardDeviation();
        Mean mean = new Mean();

        int i = 0;
        for (Double tmpData: sumDf){
            sumDfDouble[i] = tmpData;
            i++;
        }

        double sumDfStd = std.evaluate(sumDfDouble);
        double sumDfMean = mean.evaluate(sumDfDouble);

        NormalDistribution normalDis = new NormalDistribution();

        if (remainTimeMax <= 0){
            probability = 1.0;
        }
        else {
            double z_score = (remainTimeMax - sumDfMean) / sumDfStd;
            probability = 1 - normalDis.cumulativeProbability(z_score);
        }

        return probability;
    }


    public static void main(String[] args) throws Exception{
        Warning test = new Warning();


//        // test initStack()
//        // 遍历优先级栈
//        Stack<Set<Integer>> propertyStack_ = test.initStack();
//        while(!propertyStack_.isEmpty()){
//            Set<Integer> tmpData = propertyStack_.pop();
//            Iterator it=tmpData.iterator();
//            while(it.hasNext()){
//                System.out.println(it.next());
//            }
//            System.out.println();
//        }


//        // test filterResult(Set<Integer> result)
//        Set<Integer> testFilter = new HashSet<Integer>(){{
//            add(5);
//            add(6);
//        }};
//
//        Set<Integer> resultSet = test.filterResult(testFilter);
//
//        for(int i : resultSet){
//            System.out.print(i);
//        }

        // test warning
        double[] test_data = {7.7, 21.86, 7.88, 22.53, Double.NaN, Double.NaN, Double.NaN, Double.NaN, Double.NaN, Double.NaN, Double.NaN, Double.NaN, 19.71};
        String start_time = "";
        int[] need_index = argsort(test_data, true);
        Stack<Set<Integer>> property_stack_ = initStack();
        Set<Integer> result = warning(need_index, property_stack_, test_data, start_time);

        for (int i : result){
            System.out.println();
            System.out.println(i);
        }

    }
}

